SQL MetaDataExporter with custom sql types

I've a use case where I would like the use the PostgreSQL full text support ('tsvector' sql type) within the context of a QueryDSL query.

In order to be able to do this I had to implement the following classes:

- querydsl.TsVector
- querydsl.TsVectorExpression (contains methods to build the expressions we need for our use cases)
- querydsl.TsVectorPath

With these classes I can create every query I need. 

(I'm currently only interested in creating custom expressions on 'tsvector' columns. Selecting, inserting or updating is not required and I'm aware how to implement this in case I do need it.) 

Obviously the code generator of QueryDSL doesn't know about these classes. In order to avoid writing my own metamodel classes by hand I investigated the possibility to have the MetaDataExporter generate classes with fields using the custom TsPath type. I came up with the following:

```java
Connection c = ...;

SQLTemplates templates = new PostgreSQLTemplates();
Configuration configuration = new Configuration(templates);
configuration.registerType("tsvector", querydsl.TsVector.class)

TypeMappings typeMappings = new JavaTypeMappings();
Type type = new ClassType(querydsl.TsVector.class);
typeMappings.register(type, new ClassType(querydsl.TsVector.class, type));

MetaDataExporter exporter = new MetaDataExporter();
exporter.setConfiguration(configuration);
exporter.setTypeMappings(typeMappings);
...
exporter.export(c.getMetaData());
```

This setup results in a metamodel that is very close to what I would like it to be. It results in fields like this:

```java
public final querydsl.TsVectorPath<querydsl.TsVector> tsv = createTsVector("tsv", querydsl.TsVector.class);
```

Unfortunately this doesn't compile because 'createTsVector' doesn't exist. While analyzing the source of the code generator I discovered that this behavior (i.e. expecting these create methods to be available for custom types) is actually hard coded (https://github.com/querydsl/querydsl/blob/master/querydsl-sql-codegen/src/main/java/com/querydsl/sql/codegen/MetaDataSerializer.java#L321).

I figured that in order to get this code compiling, all I had to do was figure out a way to include an 'import static' that brings an appropriate 'creatTsVector' method in scope. The MetaDataExporter doesn't have a setting to include an additional 'import static' but it does have one to include an ordinary 'import' that can be used to trick the code generator into generating a 'static import':

```java
exporter.setImports(new String[]{"static querydsl.factory"});
```

results in:

```java
import static querydsl.factory.*;
```

Which does exactly what I need, provided there is a class named 'querydsl.factory' (with lowercase 'f' to trick the code generator into thinking that I want to import a package, resulting in '.*' to be added) with:

```java
public static <T extends querydsl.TsVector> querydsl.TsVectorPath<T> createTsVector(String var, Class<T> type) {
...
}
```

Although this solution works very well, it does feel like an ugly hack. I've been looking for a better strategy but everything I came up with (e.g. implementing my own Serializer, extending MetaDataExporter or creating a SQLCodegenModule.ENTITYPATH_TYPE like was done to implement spatial support) required a significant amount of additional work.

It would be great if support for using custom database types in the code generator would be easier and fully supported. It would enable me to leverage the power of specialized sql types (there are quite a few of those in the PostgreSQL universe!) without loosing the ability to use QueryDSL to create my queries.

Hi @copierrj ,

Thanks for the examples. However I haven't manged to get it working. 

My setup is like fallows:
````
dependencies {
    queryDSLMetaDataExporter project(':commons')
    queryDSLMetaDataExporter sourceSets.main.output
    queryDSLMetaDataExporter "com.querydsl:querydsl-sql:$queryDslVersion"
    queryDSLMetaDataExporter "com.querydsl:querydsl-sql-codegen:$queryDslVersion"
    queryDSLMetaDataExporter "org.postgresql:postgresql:${postgresqlVersion}"
    queryDSLMetaDataExporter "org.slf4j:slf4j-simple:${slf4jVersion}"

    compile project(':commons')
}

task querydslSqlMetadataExport(dependsOn: configurations.queryDSLMetaDataExporter.buildDependencies)

querydslSqlMetadataExport << {
    URLClassLoader loader = GroovyObject.class.getClassLoader()
    configurations.queryDSLMetaDataExporter.each { File file -> loader.addURL(file.toURI().toURL()) }

    SQLTemplates templates = new PostgreSQLTemplates();
    Configuration configuration = new Configuration(templates)

    def communicationType = Class.forName('gr8pi.web.platform.CommunicationType', true, loader) // line 112
    def communicationTypeType = Class.forName('gr8pi.querydsl.CommunicationTypeType', true, loader) // line 113

    configuration.register(communicationTypeType.newInstance())
    configuration.register('communication_tmpls', 'type', communicationType) // line 115

    TypeMappings typeMappings = new JavaTypeMappings()
    Type type = new ClassType(communicationType)
    typeMappings.register(type, new ClassType(communicationTypeType, type))

    def exporter = new MetaDataExporter()

    exporter.setImports(['gr8pi.web.platform', 'gr8pi.querydsl'] as String[]);
    exporter.setPackageName("gr8pi.platform.sql")
    exporter.setTargetFolder(new File(queryDslGenerated))
    exporter.setConfiguration(configuration)
    exporter.setTypeMappings(typeMappings)
....
}
````

And I get 
````
	at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:127)
	at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
Caused by: groovy.lang.MissingMethodException: No signature of method: com.querydsl.sql.Configuration.register() is applicable for argument types: (gr8pi.querydsl.CommunicationTypeType) values: [gr8pi.querydsl.CommunicationTypeType@5add0cc0]
Possible solutions: register(com.querydsl.sql.types.Type), register(java.lang.String, java.lang.String, com.querydsl.sql.types.Type), register(java.lang.String, java.lang.String, java.lang.Class)
	at build_5o01qqbc6lqg6adfonw1vesce$_run_closure6.doCall(//gr8pi/database/build.gradle:115)
````

I believe the issue is because I'm loading CommunicationTypeType class with another class loader. But If I use *Class.forName('path.to.class')* I get: 

````
	at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:127)
	at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
Caused by: java.lang.ClassNotFoundException: gr8pi.web.platform.CommunicationType
	at build_5o01qqbc6lqg6adfonw1vesce$_run_closure6.doCall(/gr8pi/database/build.gradle:112)
````

I'm using gradle 2.11. Any help or working example would be greatly appreciated. I'm slowly loosing my mind over this :( .

--- a/querydsl-sql-codegen/src/main/java/com/querydsl/sql/codegen/MetaDataSerializer.java
+++ b/querydsl-sql-codegen/src/main/java/com/querydsl/sql/codegen/MetaDataSerializer.java
@@ -321,11 +321,15 @@ public class MetaDataSerializer extends EntitySerializer {
 
     @Override
     protected void customField(EntityType model, Property field, SerializerConfig config,
-            CodeWriter writer) throws IOException {
+                               CodeWriter writer) throws IOException {
         Type queryType = typeMappings.getPathType(field.getType(), model, false);
-        String localRawName = writer.getRawName(field.getType());
-        serialize(model, field, queryType, writer, "create" + field.getType().getSimpleName(),
-                writer.getClassConstant(localRawName));
+        if (queryType.getPackageName().startsWith("com.querydsl")) {
+            String localRawName = writer.getRawName(field.getType());
+            serialize(model, field, queryType, writer, "create" + field.getType().getSimpleName(),
+                    writer.getClassConstant(localRawName));
+        } else {
+            super.customField(model, field, config, writer);
+        }
     }
 
     protected void serializePrimaryKeys(EntityType model, CodeWriter writer,

--- a/querydsl-sql-codegen/src/test/java/com/querydsl/sql/codegen/MetaDataSerializerTest.java
+++ b/querydsl-sql-codegen/src/test/java/com/querydsl/sql/codegen/MetaDataSerializerTest.java
@@ -13,6 +13,8 @@
  */
 package com.querydsl.sql.codegen;
 
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.util.Set;
 
@@ -27,8 +29,12 @@ import org.junit.rules.TemporaryFolder;
 import com.mysema.codegen.SimpleCompiler;
 import com.querydsl.codegen.BeanSerializer;
 import com.querydsl.sql.AbstractJDBCTest;
+import com.querydsl.sql.Configuration;
+import com.querydsl.sql.SQLTemplates;
+import com.querydsl.sql.types.AbstractType;
 
 public class MetaDataSerializerTest extends AbstractJDBCTest {
+    public static class CustomNumber { }
 
     @Rule
     public TemporaryFolder folder = new TemporaryFolder();
@@ -41,6 +47,7 @@ public class MetaDataSerializerTest extends AbstractJDBCTest {
         statement.execute("drop table survey if exists");
         statement.execute("drop table date_test if exists");
         statement.execute("drop table date_time_test if exists");
+        statement.execute("drop table spaces if exists");
 
         // survey
         statement.execute("create table survey (id int, name varchar(30), "
@@ -83,6 +90,40 @@ public class MetaDataSerializerTest extends AbstractJDBCTest {
         exporter.setPackageName("test");
         exporter.setTargetFolder(folder.getRoot());
         exporter.setNamingStrategy(namingStrategy);
+        exporter.export(connection.getMetaData());
+
+        compile(exporter);
+    }
+
+    @Test
+    public void customized_serialization() throws SQLException {
+        String namePrefix = "Q";
+        Configuration conf = new Configuration(SQLTemplates.DEFAULT);
+        conf.register("EMPLOYEE", "ID", new AbstractType<CustomNumber>(0) {
+            @Override
+            public Class<CustomNumber> getReturnedClass() {
+                return CustomNumber.class;
+            }
+
+            @Override
+            public CustomNumber getValue(ResultSet rs, int startIndex) throws SQLException {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public void setValue(PreparedStatement st, int startIndex, CustomNumber value) throws SQLException {
+                throw new UnsupportedOperationException();
+            }
+        });
+        NamingStrategy namingStrategy = new DefaultNamingStrategy();
+        // customization of serialization
+        MetaDataExporter exporter = new MetaDataExporter();
+        exporter.setBeanSerializerClass(BeanSerializer.class);
+        exporter.setNamePrefix(namePrefix);
+        exporter.setPackageName("test");
+        exporter.setTargetFolder(folder.getRoot());
+        exporter.setNamingStrategy(namingStrategy);
+        exporter.setConfiguration(conf);
         exporter.export(connection.getMetaData());
 
         compile(exporter);

